### PR TITLE
Improve checkcast/instanceof performance for interfaces on x86/x64

### PR DESCRIPTION
**Pull request by 0xdaryl**
PR #2361 introduced a number of JIT performance enhancements for checkcasts
involving interface classes.  Two of the most significant changes were the
introduction of a one-slot, dynamic cache to cache the last successful instance
class that matched the cast class.  It also inlined the itable walk to determine
if the interface class implemented the cast class.

Each time a successful checkcast was performed the cache would be updated with
that result at runtime (it was an LRU cache).  This cache was implemented as a
single, 8-byte data snippet located in the code cache.  The performance of this
approach may be acceptable for checkcast sites that do not see more than one
instance class, or when there is only a single thread of execution along this
path.  However, in the presence of multiple threads and multiple instance classes
the performance penalty of multiple threads writing to the same data address (the
class cache) is quite significant due to hardware cache coherency protocols.
In fact, the performance overhead gets significantly worse the more threads that
are involved.

There is no information provided in #2361 for what motivated the change or the
workload it was expected to benefit.

This PR modifies that implementation as follows:

* Consult available profiling information to determine how many instance classes
  may be seen by this checkcast/instanceof site.  If there are more than one then
  do not use a cache.  If there is no information available then do not use a
  cache as the characteristics are unknown.

* If there is a single profiled instance class then pre-populate the cache with
  that class.  Do not update the cache at runtime.

* If a cache is not used, then inline the interface table walk in mainline code
  rather than from outlined instructions.

Three environment variables have been introduced to control behaviour of this
evaluator:

1) `TR_updateInterfaceCheckCastCacheSlot` : when set the interface cast cache
   slot will be updated at runtime (i.e., this is the original behaviour)

2) `TR_forceDisableInterfaceCastCache` : never use the cache regardless of
   profiling information.

3) `TR_forceEnableInterfaceCastCache` : force the use of the cache regardless of
   profiling information.

2) and 3) are mutually exclusive.  Behaviour is undefined if both are set.

Using 1) and 3) will restore original behaviour.

Signed-off-by: Daryl Maier <maier@ca.ibm.com>

### CODE REVIEW

**Comment by 0xdaryl** [comment]
Jenkins test sanity.functional,extended.functional xlinux,win,osx jdk17

**Comment by 0xdaryl** [comment]
Jenkins test sanity.functional,extended.functional win jdk17

**Comment by 0xdaryl** [comment]
@BradleyWood @a7ehuo : would you mind reviewing this PR please?

FYI @JamesKingdon

**Comment by 0xdaryl** [comment]
This was also tested internally within IBM on 32-bit Windows JDK8.

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp [hunk 1]
```diff
@@ -3968,160 +3968,324 @@ inline void generateInlinedCheckCastForDynamicCastClass(TR::Node* node, TR::Code
...
-      // Not found
-      generateVFPRestoreInstruction(vfp, node, cg);
-      generateLabelInstruction(TR::InstOpCode::label, node, iTableLookUpFailLabel, cg);
-      if (isCheckCast)
+         // Fall through to endLabel
+         }
+      else
          {
-         if (tmp)
+         TR_ASSERT_FATAL(scratchReg, "Scratch register required for iTable lookup");
```
Comment (a7ehuo) [inline runtime/compiler/x/codegen/J9TreeEvaluator.cpp:4251]: If scratch register is required for iTable lookup, shouldn't this fatal assert be placed before obtaining iTable at line 4226? And `scratchReg` doesn't seem to be used after this line.

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp [hunk 2]
```diff
@@ -3968,160 +3968,324 @@ inline void generateInlinedCheckCastForDynamicCastClass(TR::Node* node, TR::Code
...
+
+         // CheckCast iTable fail lookup out-of-line
+         TR_OutlinedInstructionsGenerator og(iTableLookUpFailLabel, node, cg);
+
+         generateRegInstruction(TR::InstOpCode::PUSHReg, node, instanceClassReg, cg);
```
Comment (a7ehuo) [inline runtime/compiler/x/codegen/J9TreeEvaluator.cpp:4256]: Why should `instanceClassReg` be pushed to the stack here when checkcast fails?

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp [hunk 3]
```diff
@@ -3968,160 +3968,324 @@ inline void generateInlinedCheckCastForDynamicCastClass(TR::Node* node, TR::Code
...
+      /**
+       * Do not use a cache.  Generate the interface lookup inline and the fail
+       * path out-of-line.
+       */
+
+      /**
+       * Use the scratch register if it is available.  Otherwise, re-use the
+       * instanceClassReg register to perform itable lookup
+       */
+      TR::Register *itableReg = scratchReg ? scratchReg : instanceClassReg;
```
Comment (a7ehuo) [inline runtime/compiler/x/codegen/J9TreeEvaluator.cpp:4224]: If reusing `instanceClassReg`, should it be pushed to the stack to be saved?

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp [hunk 4]
```diff
@@ -3968,160 +3968,324 @@ inline void generateInlinedCheckCastForDynamicCastClass(TR::Node* node, TR::Code
...
-         }
-      generateLabelInstruction(TR::InstOpCode::JMP4, node, endLabel, cg);
 
-      // Not found
-      generateVFPRestoreInstruction(vfp, node, cg);
-      generateLabelInstruction(TR::InstOpCode::label, node, iTableLookUpFailLabel, cg);
-      if (isCheckCast)
+         // Fall through to endLabel
+         }
+      else
```
Comment (a7ehuo) [inline runtime/compiler/x/codegen/J9TreeEvaluator.cpp:4249]: I wonder if the case (`not found && !isCheckCast`) is missing here. 

```
      if (!isCheckCast)
         {
         // Class found in itable
         generateInstruction(TR::InstOpCode::STC, node, cg);
         // Fall through to endLabel
         }
      else
         { 
         // isCheckCast && not found 
         // What about (!isCheckCast && not found)?
          ...
         TR_OutlinedInstructionsGenerator og(iTableLookUpFailLabel, node, cg);
         ... 
         }
```

I wonder if the logic might be like this? 
```
      if (!isCheckCast)
         {
         // Class found in itable
         generateInstruction(TR::InstOpCode::STC, node, cg);

         // Fall through to endLabel
         }

      // iTable fail lookup out-of-line
      TR_OutlinedInstructionsGenerator og(iTableLookUpFailLabel, node, cg);
      if (isCheckCast)
         { .... }
      else
          { ... } 

      og.endOutlinedInstructionSequence();
```

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp [hunk 5]
```diff
@@ -3968,160 +3968,324 @@ inline void generateInlinedCheckCastForDynamicCastClass(TR::Node* node, TR::Code
...
+    * using a cache could cause performance problems as multiple threads compete to
+    * update the single cached class.  If no successful class checks have been seen
+    * before then do not speculate that a cache would be helpful.
+    */
+   bool doClassCache = (numSuccessfulClassChecks == 1) ? true : false;
+
+   static bool disableInterfaceCastCache = feGetEnv("TR_forceDisableInterfaceCastCache") != NULL;
+   if (disableInterfaceCastCache) { doClassCache = false; }
+   static bool enableInterfaceCastCache = feGetEnv("TR_forceEnableInterfaceCastCache") != NULL;
+   if (enableInterfaceCastCache) { doClassCache = true; }
```
[the remark this answers is not in the document]
Comment (BradleyWood) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp:4072]: This expression could be simplified as `enableInterfaceCastCache || doClassCache`

**Review by 0xdaryl — COMMENTED** [review]
Comments addressed.  Fixes in forced push.

**Comment by 0xdaryl** [comment]
Jenkins test sanity.functional,extended.functional xlinux,win,osx jdk17

**Comment by 0xdaryl** [comment]
Jenkins test sanity.functional,extended.functional xlinux,win,osx jdk17

**Comment by 0xdaryl** [comment]
Jenkins test extended.functional xlinux jdk17

**Comment by a7ehuo** [comment]
LGTM. I'm not able to open this PR in browsers. I got an error "This page is taking too long to load". I have no issues to open other OpenJ9 PRs. I reviewed the update in my iPad GitHub app.

**Comment by BradleyWood** [comment]
LGTM

**Review by ymanton — APPROVED** [review]
I have to look at this via the GH app on my phone, so I've only done a very cursory review.

**Pull request merged**